### ISSUE TITLE
[Feature] Ignore duplicate edges while building the subgraph for the job queue 

### ISSUE THREAD

**Issue by MatthieuBlais** [dbt-labs/dbt-core#7191]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Context: I'm investigating an issue with the command `dbt build` that became very slow in our project (1,300+ models and 20,000+ tests). It impacts user productivity (waiting time) and time-sensitive scheduled jobs. 

I suspect the function [Graph.get_subset_graph](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/graph/graph.py#L68) can become slow when running `dbt build` in a project with wide tables and many tests.

Let's consider a simple example:
```
Source A -----> Model B --------> Model C
     |                           |                               |
    \/                          \/                              \/ 
  Test A                Test B                     Test C
```

The [function](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/graph/graph.py#L68) loops through all the graph nodes and extract the upstream & downstream nodes. It creates the product of these nodes and adds the edges to the subgraph.
```
for node in self:
         if node not in include_nodes:
                source_nodes = [x for x, _ in new_graph.in_edges(node)]
                target_nodes = [x for _, x in new_graph.out_edges(node)]

                new_edges = product(source_nodes, target_nodes)
                non_cyclic_new_edges = [
                    (source, target) for source, target in new_edges if source != target
                ]  # removes cyclic refs

                new_graph.add_edges_from(non_cyclic_new_edges)
                new_graph.remove_node(node)
```

During `dbt build`, when evaluating the node `Test B` (and looking at the test nodes only), `source_nodes = [Test A]` and `target_nodes = [Test C]`. So, the edges to add to the subgraph are `[(Test A, Test C)]`. 
(_ -> Are these dependencies related to [add_test_edges](https://github.com/dbt-labs/dbt-core/blob/4186f99b742b47e0e95aca4f604cc09e5c67a449/core/dbt/compilation.py#L427)?_ ) 

Now, let's consider we have 30 tests in `Source A` and 500 tests in `Model C` --> We get 15000 new edges to add to the graph for node `Test B`.
And now let's consider we also have 100 tests in `Model B`. For all of them, we have the same source nodes and target nodes, so we have the same 15000 edges to add 100 times to the new graph (1,500,000 edges). The high number of edges to add makes the function `new_graph.add_edges_from` slow.

As the edges are 100 times the same, I think we could prevent duplicate work by ignoring edges that already exist in the new graph.

```
new_edges = [
     (source, target) for source, target in new_edges if source != target and not new_graph.has_edge(source, target)
]  # removes cyclic refs and edges already existing in the new graph
```

What do you think? 

Note that the issue doesn't seem to happen if we run `dbt run` or `dbt test` (because `source_nodes` is empty in these cases). 

### Describe alternatives you've considered

I don't observe the same behavior when running `dbt run` or `dbt test`. So, alternatively, we can use these 2 commands sequentially instead of `dbt build`, but it's not very friendly.

### Who will this benefit?

`dbt build ...` is a standard CLI command many DBT users execute.

### Are you interested in contributing this feature?

Yes

**Comment by ttusing** [comment]
Added https://github.com/dbt-labs/dbt-core/issues/7195 and got my dev environment up/PR, excited for multiple possible efficiency gains on this!

**Comment by MatthieuBlais** [comment]
Thanks for linking the issues @ttusing , I had not seen them. 

I think the idea of trimming the graph is the way to go as it is scalable. Preventing duplicate edges alone doesn't prevent the runtime for `dbt build` to slowly increase over time as the number of nodes increase. I don't know enough about the source code to evaluate the impact of your change though, so I leave it to the maintainers. 

I've tested your changes though and it looks promising!

`dbt build` runtime in my project:
- Current: ~10min
- With this PR only: ~3min
- With this PR and your PR: ~1:45min
- With your PR only: ~1:45min

So I think if your change doesn't have any side effect, we could keep yours only!

**Comment by ttusing** [comment]
> Thanks for linking the issues @ttusing , I had not seen them.
> 
> I think the idea of trimming the graph is the way to go as it is scalable. Preventing duplicate edges alone doesn't prevent the runtime for `dbt build` to slowly increase over time as the number of nodes increase. I don't know enough about the source code to evaluate the impact of your change though, so I leave it to the maintainers.
> 
> I've tested your changes though and it looks promising!
> 
> `dbt build` runtime in my project:
> 
> * Current: ~10min
> * With this PR only: ~3min
> * With this PR and your PR: ~1:45min
> * With your PR only: ~1:45min
> 
> So I think if your change doesn't have any side effect, we could keep yours only!

I think your changes can still improve runtime for some selectors that still require adding edges after removing nodes. For instance, if you have two disconnected subgraphs that are connected in the main DAG, or a model that is re-joined multiple times downstream, this could shave off a few seconds I'd bet.

Although to be fair I am still unsure why the subgraph needs to calculate these distant ancestors in the first place 😅 . Seems like the type of selector comes up uncommon enough from my usage patterns that I'm not too concerned in picking it apart.

**Comment by ttusing** [comment]
@MatthieuBlais I added an adjustment to my PR to sort the DAG by node degree before searching and it went down to 10 seconds build time (down from 26 in last commit)! Hopefully this is also helpful for you.

https://github.com/dbt-labs/dbt-core/pull/7194

**Comment by MatthieuBlais** [comment]
> I think your changes can still improve runtime for some selectors that still require adding edges after removing nodes. For instance, if you have two disconnected subgraphs that are connected in the main DAG, or a model that is re-joined multiple times downstream, this could shave off a few seconds I'd bet.

Yes you're right, after trimming the graph, the model selector matters. I can see a few additional seconds being shaved off. With your change, now the waiting time (excluding partial parsing time) is down to a few seconds only, that looks awesome!

**Comment by jtcohen6** [comment]
@ttusing @MatthieuBlais Awesome conversation! Would it make sense to consolidate this issue with #7195, as well as the two associated PRs (#7192 + #7194)? Or would it still make sense to consider & review them independently?

**Comment by peterallenwebb** [comment]
@ttusing @MatthieuBlais Thanks so much for your high-quality analysis and coding on this. Our subject area expert still needs to weigh in on the PRs, but they look good to me. We've been aware that this code was bottlenecking some use cases, but to my knowledge the Core team hasn't had access to real-world dbt projects that exhibited the issue in a way that has enabled us to optimize it. With that in mind, if you are willing and able to share your projects with us, it would be enormously helpful as we monitor and improve performance in real-world scenarios. I can arrange a private channel for that if it's something you can do.

**Comment by ttusing** [comment]
@peterallenwebb Yeah, sounds tricky. Think our legal function would want to look at this.

We use DBT cloud, is that potentially a path?

It might also be helpful to establish a pool of testers to run their own benchmarks and report back on a variety of projects, connectors, machines, envs, etc. Does such a thing already exist?

**Comment by boxysean** [comment]
@ttusing, in support of @peterallenwebb's request, I have a little script [here](https://gist.github.com/boxysean/1a502e431beb8ed8b35f9652987ceabc#file-generate_model_folder-py) that takes in your `manifest.json` file and outputs a dbt models folder with the same DAG shape, but all models are simply `SELECT 1`. This project could then be run using (say) dbt-duckdb, for profiling purposes.

Would you be willing to share the output of the script?

**Comment by ttusing** [comment]
I think so, let me run it by our team.
